File: app/src/main/java/space/pal/sig/repository/NewsRepository.java
package space.pal.sig.repository;

import androidx.lifecycle.LiveData;
import androidx.paging.LivePagedListBuilder;
import androidx.paging.PagedList;

import java.util.List;

import io.reactivex.Completable;
import lombok.RequiredArgsConstructor;
import space.pal.sig.model.News;
import space.pal.sig.model.NewsSource;
import space.pal.sig.model.database.NewsDao;

/**
 * SpaceNow
 * Created by Catalin on 7/18/2020
 **/
@RequiredArgsConstructor
public class NewsRepository {

    private final NewsDao newsDao;

    public Completable insertOrUpdate(List<News> newsList) {
        return newsDao.insert(newsList.toArray(new News[0]));
    }

    public Completable insertOrUpdate(News news) {
        return newsDao.insert(news);
    }

    public LiveData<PagedList<News>> findAllBySource(NewsSource source, int pageSize) {
        return new LivePagedListBuilder<>(newsDao.findAllBySource(source), pageSize).build();
    }

    public LiveData<News> findById(String id) {
        return newsDao.findById(id);
    }

}

File: app/src/main/java/space/pal/sig/repository/NewsRepository.kt
package space.pal.sig.repository

import androidx.lifecycle.LiveData
import io.reactivex.Single
import space.pal.sig.database.dao.NewsDao
import space.pal.sig.model.dto.NewsDto
import space.pal.sig.model.dto.NewsPreviewDto
import space.pal.sig.model.entity.News
import space.pal.sig.network.HubbleService

/**
 * SpaceNow
 * Created by Catalin on 12/25/2020
 **/
class NewsRepository(
        private val newsDao: NewsDao,
        private val hubbleService: HubbleService
) {

    fun save(news: News) {
        newsDao.save(news)
    }

    fun findAll(): LiveData<MutableList<News>> {
        return newsDao.findAll()
    }

    fun downloadHubbleFeed(page: Int): Single<List<NewsPreviewDto>> {
        return hubbleService.getHubbleFeed(page)
    }

    fun downloadHubbleNews(newsId: String): Single<NewsDto> {
        return hubbleService.getHubbleNews(newsId)
    }

    fun downloadEuropeanSpaceAgencyFeed(page: Int): Single<List<NewsDto>> {
        return hubbleService.getEuropeanSpaceAgencyFeed(page)
    }

    fun downloadJamesWebbSpaceTelescopeFeed(page: Int): Single<List<NewsDto>> {
        return hubbleService.getJamesWebbSpaceTelescopeFeed(page)
    }

    fun downloadSpaceTelescopeLiveFeed(page: Int): Single<List<NewsDto>> {
        return hubbleService.getSpaceTelescopeLiveFeed(page)
    }

}